Simplify JavaScriptEvaluationResult extraction and insertion

achristensen07 opened the pull request:
#### 2abf1685b31248d3c8f90448c83cbda3ddd718a2
<pre>
Simplify JavaScriptEvaluationResult extraction and insertion
<a href="https://bugs.webkit.org/show_bug.cgi?id=297490">https://bugs.webkit.org/show_bug.cgi?id=297490</a>
<a href="https://rdar.apple.com/158441459">rdar://158441459</a>

Reviewed by Sihui Liu.

JavaScriptEvaluationResult is an IPC-friendly representation of serializable JavaScript
values that can be extracted from and inserted into four different API environments:
JavaScript, GLib, ObjC, and WebKit&apos;s API abstractions.  This simplifies and cleans up
those transitions.  Not only does it reduce sizeof(JavaScriptEvaluationResult) and move
the structures associated with transitions onto the stack with the correct lifetimes,
but it also cleans up layering a bit and isolates the handling of null in those environments,
removing the need for m_nullObjectID and replacing it with JavaScriptEvaluationResult::jsUndefined.

This is a necessary step before replacing roundTripThroughSerializedScriptValue with
a faster solution that will also be able to handle JSWebKitJSHandle and JSWebKitSerializedNode
objects inside arrays and maps.

This functionality is well covered by many tests, including TestWebKitAPI.EvaluateJavaScript.ReturnTypes,
TestWebKitAPI.WKWebView.EvaluateJavaScriptErrorCases, the EvaluateJavaScript API tests, and many others.

* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::APIInsertionContext::toAPI):
(WebKit::JavaScriptEvaluationResult::APIExtractionContext::toValue):
(WebKit::JavaScriptEvaluationResult::extract):
(WebKit::JavaScriptEvaluationResult::APIExtractionContext::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::toAPI):
(WebKit::JavaScriptEvaluationResult::JSExtractionContext::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::JSExtractionContext::toValue):
(WebKit::JavaScriptEvaluationResult::JSInsertionContext::toJS):
(WebKit::JavaScriptEvaluationResult::toJS):
(WebKit::JavaScriptEvaluationResult::jsUndefined):
(WebKit::JavaScriptEvaluationResult::toValue): Deleted.
(WebKit::JavaScriptEvaluationResult::addObjectToMap): Deleted.
(WebKit::JavaScriptEvaluationResult::toWK): Deleted.
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
(WebKit::JavaScriptEvaluationResult::map const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::ObjCInsertionContext::toID):
(WebKit::JavaScriptEvaluationResult::toID):
(WebKit::JavaScriptEvaluationResult::ObjCExtractionContext::toValue):
(WebKit::JavaScriptEvaluationResult::ObjCExtractionContext::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::extract):
(WebKit::JavaScriptEvaluationResult::toValue): Deleted.
(WebKit::JavaScriptEvaluationResult::addObjectToMap): Deleted.
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult): Deleted.
* Source/WebKit/Shared/glib/JavaScriptEvaluationResultGLib.cpp:
(WebKit::JavaScriptEvaluationResult::GLibExtractionContext::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::GLibExtractionContext::toValue):
(WebKit::JavaScriptEvaluationResult::extract):
(WebKit::JavaScriptEvaluationResult::addObjectToMap): Deleted.
(WebKit::JavaScriptEvaluationResult::toValue): Deleted.
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult): Deleted.
* Source/WebKit/UIProcess/API/APIScriptMessage.cpp:
(API::ScriptMessage::ScriptMessage):
* Source/WebKit/UIProcess/API/APIScriptMessage.h:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageEvaluateJavaScriptInFrame):
(WKPageCallAsyncJavaScript):
* Source/WebKit/UIProcess/API/C/WKScriptMessageRef.cpp:
(WKScriptMessageGetBody):
* Source/WebKit/UIProcess/API/C/WKUserContentControllerRef.cpp:

Canonical link: <a href="https://commits.webkit.org/298845@main">https://commits.webkit.org/298845@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9f56703131f4e5144b1e33886621bdf3c511fdfd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116850 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36515 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27088 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122935 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67440 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d1dacf6d-a086-4e8c-a887-a03d21c8ed60) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37212 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45115 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88737 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43144 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e2d514b7-53ff-4a95-b666-942b841d655f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119799 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29675 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104821 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69197 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28739 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66591 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99055 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23080 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126062 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43750 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32868 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97406 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44113 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101021 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97204 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42533 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20472 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39766 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18655 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43635 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49231 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43102 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46441 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44807 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/Shared/JavaScriptEvaluationResult.h:
```diff
@@ -85,65 +86,72 @@ class JavaScriptEvaluationResult {
-    Value toValue(GVariant*);
-    JSObjectID addObjectToMap(GVariant*);
-#endif
+    struct GLibExtractionContext {
```
Can we make `GLibExtractionContext` and other context `class` and their members private? It is a bit hard to tell whether their functions are modifying struct member or a local variable (e.g. `map` in `toValue` function).

achristensen07 committed:
Simplify JavaScriptEvaluationResult extraction and insertion
https://bugs.webkit.org/show_bug.cgi?id=297490
rdar://158441459

Reviewed by Sihui Liu.

JavaScriptEvaluationResult is an IPC-friendly representation of serializable JavaScript
values that can be extracted from and inserted into four different API environments:
JavaScript, GLib, ObjC, and WebKit's API abstractions.  This simplifies and cleans up
those transitions.  Not only does it reduce sizeof(JavaScriptEvaluationResult) and move
the structures associated with transitions onto the stack with the correct lifetimes,
but it also cleans up layering a bit and isolates the handling of null in those environments,
removing the need for m_nullObjectID and replacing it with JavaScriptEvaluationResult::jsUndefined.

This is a necessary step before replacing roundTripThroughSerializedScriptValue with
a faster solution that will also be able to handle JSWebKitJSHandle and JSWebKitSerializedNode
objects inside arrays and maps.

This functionality is well covered by many tests, including TestWebKitAPI.EvaluateJavaScript.ReturnTypes,
TestWebKitAPI.WKWebView.EvaluateJavaScriptErrorCases, the EvaluateJavaScript API tests, and many others.

* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult):
(WebKit::JavaScriptEvaluationResult::APIInsertionContext::toAPI):
(WebKit::JavaScriptEvaluationResult::APIExtractionContext::toValue):
(WebKit::JavaScriptEvaluationResult::extract):
(WebKit::JavaScriptEvaluationResult::APIExtractionContext::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::toAPI):
(WebKit::JavaScriptEvaluationResult::JSExtractionContext::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::JSExtractionContext::toValue):
(WebKit::JavaScriptEvaluationResult::JSInsertionContext::toJS):
(WebKit::JavaScriptEvaluationResult::toJS):
(WebKit::JavaScriptEvaluationResult::jsUndefined):
(WebKit::JavaScriptEvaluationResult::toValue): Deleted.
(WebKit::JavaScriptEvaluationResult::addObjectToMap): Deleted.
(WebKit::JavaScriptEvaluationResult::toWK): Deleted.
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
(WebKit::JavaScriptEvaluationResult::map const):
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::ObjCInsertionContext::toID):
(WebKit::JavaScriptEvaluationResult::toID):
(WebKit::JavaScriptEvaluationResult::ObjCExtractionContext::toValue):
(WebKit::JavaScriptEvaluationResult::ObjCExtractionContext::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::extract):
(WebKit::JavaScriptEvaluationResult::toValue): Deleted.
(WebKit::JavaScriptEvaluationResult::addObjectToMap): Deleted.
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult): Deleted.
* Source/WebKit/Shared/glib/JavaScriptEvaluationResultGLib.cpp:
(WebKit::JavaScriptEvaluationResult::GLibExtractionContext::addObjectToMap):
(WebKit::JavaScriptEvaluationResult::GLibExtractionContext::toValue):
(WebKit::JavaScriptEvaluationResult::extract):
(WebKit::JavaScriptEvaluationResult::addObjectToMap): Deleted.
(WebKit::JavaScriptEvaluationResult::toValue): Deleted.
(WebKit::JavaScriptEvaluationResult::JavaScriptEvaluationResult): Deleted.
* Source/WebKit/UIProcess/API/APIScriptMessage.cpp:
(API::ScriptMessage::ScriptMessage):
* Source/WebKit/UIProcess/API/APIScriptMessage.h:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageEvaluateJavaScriptInFrame):
(WKPageCallAsyncJavaScript):
* Source/WebKit/UIProcess/API/C/WKScriptMessageRef.cpp:
(WKScriptMessageGetBody):
* Source/WebKit/UIProcess/API/C/WKUserContentControllerRef.cpp:

Canonical link: https://commits.webkit.org/298845@main
